### PR TITLE
XD-2894: Trigger metadata (and others) enhancements

### PR DESCRIPTION
**Pull request by ericbottard**
Introduce @DateFormat and @CronExpression jsr303 annotations.
Apply where appropriate

### CODE REVIEW

**Comment by ericbottard** [comment]
Rebased

**Comment by ghillert** [comment]
This might be a different Jira, but I thing we should handle the module options validation more gracefully. We should avoid triggering a 500 HTTP status:

````
spring> stream create vvv2 --definition "trigger --cron='123' | log"
15:34:11,294  WARN Spring Shell client.RestTemplate - POST request for "http://localhost:9393/streams/definitions" resulted in 500 (Server Error); invoking error handler

Command failed org.springframework.xd.rest.client.impl.SpringXDException: Error with option(s) for module trigger of type source:
    cron: Cron expression must consist of 6 fields (found 1 in "123")

spring> 
````

This also triggers a rather long server-side stack trace:

````
15:34:11,270 ERROR qtp528042460-40 rest.RestControllerAdvice - Caught exception while handling a request
org.springframework.xd.dirt.plugins.ModuleConfigurationException: Error with option(s) for module trigger of type source:
    cron: Cron expression must consist of 6 fields (found 1 in "123")
	at org.springframework.xd.dirt.plugins.ModuleConfigurationException.fromBindException(ModuleConfigurationException.java:55)
	at org.springframework.xd.dirt.stream.XDStreamParser.parse(XDStreamParser.java:180)
	at org.springframework.xd.dirt.stream.AbstractDeployer.validateBeforeSave(AbstractDeployer.java:116)
	at org.springframework.xd.dirt.rest.XDController.save(XDController.java:260)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:483)
	at org.springframework.web.method.support.InvocableHandlerMethod.doInvoke(InvocableHandlerMethod.java:221)
	at org.springframework.web.method.support.InvocableHandlerMethod.invokeForRequest(InvocableHandlerMethod.java:137)
	at org.springframework.web.servlet.mvc.method.annotation.ServletInvocableHandlerMethod.invokeAndHandle(ServletInvocableHandlerMethod.java:110)
	at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.invokeHandleMethod(RequestMappingHandlerAdapter.java:777)
	at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.handleInternal(RequestMappingHandlerAdapter.java:706)
	at org.springframework.web.servlet.mvc.method.AbstractHandlerMethodAdapter.handle(AbstractHandlerMethodAdapter.java:85)
	at org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:943)
	at org.springframework.web.servlet.DispatcherServlet.doService(DispatcherServlet.java:877)
	at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:966)
	at org.springframework.web.servlet.FrameworkServlet.doPost(FrameworkServlet.java:868)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:644)
	at org.springframework.web.servlet.FrameworkServlet.service(FrameworkServlet.java:842)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:725)
	at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:684)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1496)
	at org.springframework.boot.actuate.autoconfigure.EndpointWebMvcAutoConfiguration$ApplicationContextHeaderFilter.doFilterInternal(EndpointWebMvcAutoConfiguration.java:291)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1467)
	at org.springframework.web.filter.HiddenHttpMethodFilter.doFilterInternal(HiddenHttpMethodFilter.java:77)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1467)
	at org.springframework.web.filter.HttpPutFormContentFilter.doFilterInternal(HttpPutFormContentFilter.java:87)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1467)
	at org.springframework.boot.actuate.trace.WebRequestTraceFilter.doFilterInternal(WebRequestTraceFilter.java:100)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1467)
	at org.springframework.security.web.FilterChainProxy.doFilterInternal(FilterChainProxy.java:186)
	at org.springframework.security.web.FilterChainProxy.doFilter(FilterChainProxy.java:160)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1467)
	at org.springframework.boot.actuate.autoconfigure.MetricFilterAutoConfiguration$MetricsFilter.doFilterInternal(MetricFilterAutoConfiguration.java:90)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1467)
	at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:499)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:137)
	at org.eclipse.jetty.security.SecurityHandler.handle(SecurityHandler.java:557)
	at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:231)
	at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1086)
	at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:428)
	at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:193)
	at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1020)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:135)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:116)
	at org.eclipse.jetty.server.Server.handle(Server.java:370)
	at org.eclipse.jetty.server.AbstractHttpConnection.handleRequest(AbstractHttpConnection.java:494)
	at org.eclipse.jetty.server.AbstractHttpConnection.content(AbstractHttpConnection.java:982)
	at org.eclipse.jetty.server.AbstractHttpConnection$RequestHandler.content(AbstractHttpConnection.java:1043)
	at org.eclipse.jetty.http.HttpParser.parseNext(HttpParser.java:865)
	at org.eclipse.jetty.http.HttpParser.parseAvailable(HttpParser.java:240)
	at org.eclipse.jetty.server.AsyncHttpConnection.handle(AsyncHttpConnection.java:82)
	at org.eclipse.jetty.io.nio.SelectChannelEndPoint.handle(SelectChannelEndPoint.java:667)
	at org.eclipse.jetty.io.nio.SelectChannelEndPoint$1.run(SelectChannelEndPoint.java:52)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:608)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$3.run(QueuedThreadPool.java:543)
	at java.lang.Thread.run(Thread.java:745)
Caused by: org.springframework.validation.BindException: org.springframework.validation.BeanPropertyBindingResult: 1 errors
Field error in object 'target' on field 'cron': rejected value [123]; codes [CronExpression.target.cron,CronExpression.cron,CronExpression.java.lang.String,CronExpression]; arguments [org.springframework.context.support.DefaultMessageSourceResolvable: codes [target.cron,cron]; arguments []; default message [cron]]; default message [Cron expression must consist of 6 fields (found 1 in "123")]
	at org.springframework.xd.module.options.PojoModuleOptionsMetadata.bindAndValidate(PojoModuleOptionsMetadata.java:205)
	at org.springframework.xd.module.options.PojoModuleOptionsMetadata.interpolate(PojoModuleOptionsMetadata.java:139)
	at org.springframework.xd.module.options.FlattenedCompositeModuleOptionsMetadata.interpolate(FlattenedCompositeModuleOptionsMetadata.java:152)
	at org.springframework.xd.module.options.FlattenedCompositeModuleOptionsMetadata.interpolate(FlattenedCompositeModuleOptionsMetadata.java:152)
	at org.springframework.xd.module.options.EnvironmentAwareModuleOptionsMetadataResolver$ModuleOptionsMetadataWithDefaults.interpolate(EnvironmentAwareModuleOptionsMetadataResolver.java:168)
	at org.springframework.xd.dirt.stream.XDStreamParser.parse(XDStreamParser.java:177)
	... 61 more
````

File: spring-xd-module-spi/src/test/java/org/springframework/xd/module/options/validation/DateFormatValidatorTests.java [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+
+	public static class Bean {
+
+		@DateFormat
+		private String format1;
+
+		@DateFormat(message = "not a valid date format")
+		private String format2;
+	}
+}
```
Comment (ghillert) [inline spring-xd-module-spi/src/test/java/org/springframework/xd/module/options/validation/DateFormatValidatorTests.java]: Add return

File: spring-xd-module-spi/src/test/java/org/springframework/xd/module/options/validation/CronValidatorTests.java [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+
+		@CronExpression
+		private String cron1;
+
+		@CronExpression(message = "something is wrong")
+		private String cron2;
+
+	}
+
+}
```
Comment (ghillert) [inline spring-xd-module-spi/src/test/java/org/springframework/xd/module/options/validation/CronValidatorTests.java]: Add return

**Comment by ghillert** [comment]
This might also be a separate Jira.

Should a provided date be validated during stream creation? 

````
stream create www --definition "trigger --date='44/44/44' | log"
````

This should theoretically fail as `44/44/44` does not fit the default date format: 

````
dateFormat    the format specifying how the 'date' should be parsed                                                   MM/dd/yy HH:mm:ss  String
````

**Comment by ericbottard** [comment]
I created https://jira.spring.io/browse/XD-3032 for the first issue, will try to come up with something in this PR for the joint validation of the date.

**Comment by ericbottard** [comment]
Updated the PR with validation for the date

**Comment by ghillert** [comment]
Will fix during merge - LGTM otherwise!

**Comment by ghillert** [comment]
Merged.

**Pull request closed without merging**